kernels/volk/volk_32f_stddev_and_mean_32f_x2.h

```diff
@@ -73,325 +73,513 @@
+    S0_acc = vfmaq_f32(S0_acc, x0_reg, f_reg);
+
+    T1_acc = vaddq_f32(T1_acc, v1_reg);
+    x1_reg = vdupq_n_f32(np1);
+    x1_reg = vmulq_f32(x1_reg, v1_reg);
+    x1_reg = vsubq_f32(x1_reg, T1_acc);
+    x1_reg = vmulq_f32(x1_reg, x1_reg);
+    S1_acc = vfmaq_f32(S1_acc, x1_reg, f_reg);
   }
-  *stddev = returnValue;
-  *mean = newMean;
-}
-#endif /* LV_HAVE_SSE4_1 */
 
+  vst1q_f32(&T[0], T0_acc);
+  vst1q_f32(&T[4], T1_acc);
+
+  vst1q_f32(&S[0], S0_acc);
+  vst1q_f32(&S[4], S1_acc);
+
+  accrue_result( S, T, 8, eigth_points);
+
+  number = eigth_points*8;
+
+  FINALIZE:
+  for (; number < num_points; number++) {
+    update_square_sum_1_val(&S[0], &T[0], &number, in_ptr);
+    T[0] += (*in_ptr++);
+  }
+
+  *stddev = sqrtf( S[0] / num_points );
+  *mean   = T[0] / num_points;
+}
+#endif /* LV_HAVE_NEON */
 
 #ifdef LV_HAVE_SSE
 #include <xmmintrin.h>
 
 static inline void
-volk_32f_stddev_and_mean_32f_x2_a_sse(float* stddev, float* mean,
+volk_32f_stddev_and_mean_32f_x2_u_sse(float* stddev, float* mean,
                                       const float* inputBuffer,
                                       unsigned int num_points)
 {
-  float returnValue = 0;
-  float newMean = 0;
-  if(num_points > 0){
-    unsigned int number = 0;
-    const unsigned int quarterPoints = num_points / 4;
-
-    const float* aPtr = inputBuffer;
-    __VOLK_ATTR_ALIGNED(16) float meanBuffer[4];
-    __VOLK_ATTR_ALIGNED(16) float squareBuffer[4];
-
-    __m128 accumulator = _mm_setzero_ps();
-    __m128 squareAccumulator = _mm_setzero_ps();
-    __m128 aVal = _mm_setzero_ps();
-    for(;number < quarterPoints; number++) {
-      aVal = _mm_load_ps(aPtr);                     // aVal = x
-      accumulator = _mm_add_ps(accumulator, aVal);  // accumulator += x
-      aVal = _mm_mul_ps(aVal, aVal);                // squareAccumulator += x^2
-      squareAccumulator = _mm_add_ps(squareAccumulator, aVal);
-      aPtr += 4;
-    }
-    _mm_store_ps(meanBuffer,accumulator); // Store the results back into the C container
-    _mm_store_ps(squareBuffer,squareAccumulator); // Store the results back into the C container
-    newMean = meanBuffer[0];
-    newMean += meanBuffer[1];
-    newMean += meanBuffer[2];
-    newMean += meanBuffer[3];
-    returnValue = squareBuffer[0];
-    returnValue += squareBuffer[1];
-    returnValue += squareBuffer[2];
-    returnValue += squareBuffer[3];
-
-    number = quarterPoints * 4;
-    for(;number < num_points; number++){
-      returnValue += (*aPtr) * (*aPtr);
-      newMean += *aPtr++;
-    }
-    newMean /= num_points;
-    returnValue /= num_points;
-    returnValue -= (newMean * newMean);
-    returnValue = sqrtf(returnValue);
+  if (num_points == 0) { return; }
+
+  const float* in_ptr = inputBuffer;
+
+  unsigned int number = 1;
+
+  __VOLK_ATTR_ALIGNED(16) float T[8] = {0.f};
+  __VOLK_ATTR_ALIGNED(16) float S[8] = {0.f};
+
+  if (num_points < 8) {   
+    T[0] = (*in_ptr++);
+    goto FINALIZE; }
```

Comment:
This is only the case if SIMD can't be used. This should be equivalent to a call to `generic`. I suggest to move this check to the top, call `generic` in case we do not have enough input and then just return.

Reply:
ok